app/forms/support/update_provider_form.rb

```diff
@@ -2,21 +2,31 @@
 
 module Support
   class UpdateProviderForm
-    attr_reader :provider
+    delegate :valid?, to: :provider
 
     def initialize(provider, attributes:)
       @provider = provider
       @attributes = attributes
+      assign_attributes
     end
 
     def save
-      @provider.assign_attributes(@attributes)
+      return false unless provider.valid?
+
+      provider.save!
+    end
```

Comment:
In related to the PR description:

> Any tips on the form would be useful. How do we want to use forms going forward?

I'd not make the save here. On this example is simple that's why I approved but if becomes more complicated this class can be convulated (and then move to a service object to handle the transaction)

Reply:
How do you propose the save should be made?

Reply:
> How do you propose the save should be made?

Sure. I would create a Policy class and a service to handle the transaction (like in Apply we have in the provider interface).

```ruby
class AccreditingProviderPolicy
    def removed_accreditation?(provider)
      provider.accrediting_provider_changed?(to: 'not_an_accredited_provider')
    end
  end

  class ProviderPersistenceService
    def initialize(provider, attributes, accrediting_policy: AccreditingProviderPolicy.new)
      @provider = provider
      @attributes = attributes
      @accrediting_policy = accrediting_policy
    end

    def save
      ActiveRecord::Base.transaction do
        provider.assign_attributes(attributes)
        apply_accrediting_policy
        provider.save!
      end
    rescue ActiveRecord::RecordInvalid => e
      handle_error(e)
      false
    end

    private

    attr_reader :provider, :attributes, :accrediting_policy

    def apply_accrediting_policy
      if accrediting_policy.removed_accreditation?(provider)
        provider.accredited_provider_number = nil
      end
    end
  end
end
```

Then you could have the call on the form or in the controller (1 or 2 below).

1. I do prefer in the controller but I can see the argument to be in the form:

```ruby
    def save
      return false unless valid?

      ProviderPersistenceService.new(provider, attributes).save
    end
```

2. In the controller:

```ruby
def update
    provider = Provider.find(params[:id])
    form = Support::UpdateProviderForm.new(provider, attributes: provider_params)

    if form.valid? && Support::ProviderPersistenceService.new(form.provider, provider_params).save
      redirect_to provider_path(provider), notice: 'Provider updated successfully.'
    else
      flash.now[:alert] = form.valid? ? 'Failed to save provider.' : 'Invalid provider data.'
      render :edit
    end
  end
```

Reply:
I like Policy objects and I think we could have more on Publish by the way. The `AccreditingProviderPolicy` can be even receive the provider on the initialize.